Source/WebCore/css/calc/CSSCalcTree+NumericIdentity.h

```diff
@@ -0,0 +1,303 @@
+/*
+ * Copyright (C) 2024 Samuel Weinig <sam@webkit.org>
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include "CSSCalcTree.h"
+#include "CSSUnits.h"
+
+namespace WebCore {
+namespace CSSCalc {
+
+// `NumericIdentity` is used to assign a densely packed unique identifier for each possible numeric type allowed in a `calc()` expression. This allows the construction of fixed size (specifically sized to `numberOfNumericIdentityTypes`) lookup tables needed in expression simplification.
+
+enum class NumericIdentity : uint8_t {
+    Number,
+    Percent,
+
+    // Canonical dimension units
+    PX,
+    DEG,
+    S,
+    HZ,
+    DPPX,
+    FR,
+
+    // Non-canonical dimension units
+    CM,
+    MM,
+    Q,
+    INCH, // NOTE: using "IN" here breaks under MSVC.
+    PT,
+    PC,
+    EM,
+    EX,
+    LH,
+    CAP,
+    CH,
+    IC,
+    RCAP,
+    RCH,
+    REM,
+    REX,
+    RIC,
+    RLH,
+    VW,
+    VH,
+    VMIN,
+    VMAX,
+    VB,
+    VI,
+    SVW,
+    SVH,
+    SVMIN,
+    SVMAX,
+    SVB,
+    SVI,
+    LVW,
+    LVH,
+    LVMIN,
+    LVMAX,
+    LVB,
+    LVI,
+    DVW,
+    DVH,
+    DVMIN,
+    DVMAX,
+    DVB,
+    DVI,
+    CQW,
+    CQH,
+    CQI,
+    CQB,
+    CQMIN,
+    CQMAX,
+    RAD,
+    GRAD,
+    TURN,
+    MS,
+    KHZ,
+    X,
+    DPI,
+    DPCM
+};
+
+constexpr uint8_t numberOfNumericIdentityTypes = std::to_underlying(NumericIdentity::DPCM) + 1;
```

Comment:
This is a C++23 feature; short of implementing this in WTF, I think we should just use a normal `static_cast<uint8_t>`.

Reply:
Is it breaking anything? I didn't see any issues with it on EWS. We are building with c++23 already I believe (@achristensen07 please correct me if I am wrong).

Reply:
We do have WTF::enumToUnderlyingType, but now that we are using C++23, we should be migrating it to std::to_underlying.  Does this cause playstation build failures?

Reply:
Everything was changed to static_cast : https://github.com/WebKit/WebKit/commit/baaae40b9bf25c5b292c1c0bda83a74977c34715